cli/src/commands/touch.rs

```diff
@@ -0,0 +1,179 @@
+        return Ok(());
+    }
+    workspace_command.check_rewritable(commit_ids.iter())?;
+
+    let mut tx = workspace_command.start_transaction();
+    let tx_description = match commit_ids.as_slice() {
+        [] => unreachable!(),
+        [commit] => format!("touch commit {}", commit.hex()),
+        [first_commit, remaining_commits @ ..] => {
+            format!(
+                "touch commit {} and {} more",
+                first_commit.hex(),
+                remaining_commits.len()
+            )
+        }
+    };
+
+    let mut num_touched = 0;
+    let mut num_reparented = 0;
+    let commit_ids_set: HashSet<_> = commit_ids.iter().cloned().collect();
+    let mut touched: HashMap<CommitId, CommitId> = HashMap::new();
+    // Even though `MutableRepo::rewrite_commit` and
+    // `MutableRepo::rebase_descendants` can handle rewriting of a commit even
+    // if it is a descendant of another commit being rewritten, using
+    // `MutableRepo::transform_descendants` prevents us from rewriting the same
+    // commit multiple times, and adding additional entries in the predecessor
+    // chain.
+    tx.repo_mut()
+        .transform_descendants(commit_ids, async |rewriter| {
+            let old_commit_id = rewriter.old_commit().id().clone();
+            let mut commit_builder = rewriter.reparent();
+            let new_parents = commit_builder
+                .parents()
+                .iter()
+                .map(|p| (touched.get(p).unwrap_or(p)))
+                .cloned()
+                .collect();
+            commit_builder = commit_builder.set_parents(new_parents);
+            if commit_ids_set.contains(&old_commit_id) {
+                let mut new_author = commit_builder.author().clone();
+                if let Some((name, email)) = args.author.clone() {
+                    new_author.name = name;
+                    new_author.email = email;
+                } else if args.update_author {
+                    new_author.name = commit_builder.committer().name.clone();
+                    new_author.email = commit_builder.committer().email.clone();
+                }
+                if args.update_author_timestamp {
+                    new_author.timestamp = commit_builder.committer().timestamp;
+                }
+                commit_builder = commit_builder.set_author(new_author);
+
+                if args.update_change_id {
+                    commit_builder = commit_builder.generate_new_change_id();
+                }
+
+                let new_commit = commit_builder.write()?;
+                touched.insert(old_commit_id.clone(), new_commit.id().clone());
+                num_touched += 1;
+            } else {
+                commit_builder.write()?;
+                num_reparented += 1;
+            }
+            Ok(())
+        })?;
+    for (old, new) in touched {
+        tx.repo_mut().set_rewritten_commit(old.clone(), new.clone());
+    }
```

Comment:
nit: maybe better to do this only when `args.update_change_id` is specified to clarify why we do set rewritten commits manually. Adding comment is also good.